fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/StorageAdapter.java

```diff
@@ -0,0 +1,389 @@
+/*
+ * StorageAdapter.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.async.hnsw;
+
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.Range;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.StreamingMode;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.linear.AffineOperator;
+import com.apple.foundationdb.linear.DoubleRealVector;
+import com.apple.foundationdb.linear.FloatRealVector;
+import com.apple.foundationdb.linear.HalfRealVector;
+import com.apple.foundationdb.linear.Quantizer;
+import com.apple.foundationdb.linear.RealVector;
+import com.apple.foundationdb.linear.VectorType;
+import com.apple.foundationdb.rabitq.EncodedRealVector;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines the contract for storing and retrieving HNSW graph data to/from a persistent store.
+ * <p>
+ * This interface provides an abstraction layer over the underlying database, handling the serialization and
+ * deserialization of HNSW graph components such as nodes, vectors, and their relationships. Implementations of this
+ * interface are responsible for managing the physical layout of data within a given {@link Subspace}.
+ * The generic type {@code N} represents the specific type of {@link NodeReference} that this storage adapter manages.
+ *
+ * @param <N> the type of {@link NodeReference} this storage adapter manages
+ */
+interface StorageAdapter<N extends NodeReference> {
+    ImmutableList<VectorType> VECTOR_TYPES = ImmutableList.copyOf(VectorType.values());
+
+    /**
+     * Subspace for the index access info; contains entry nodes; these are kept separately from the data.
+     */
+    byte SUBSPACE_PREFIX_INDEX_ACCESS_INFO = 0x00;
+
+    /**
+     * Subspace for data.
+     */
+    byte SUBSPACE_PREFIX_DATA = 0x01;
```

Comment:
I'm not quite sure how much this matters, but the encoding for 1 is two bytes long (`\x15\x01`), whereas the encoding for 0 is only one byte (`\x14`). We may be able to keep the keys slightly shorter if we make `SUBSPACE_PREFIX_DATA` be zero instead of `SUBSPACE_PREFIX_INDEX_ACCESS_INFO`. The value `null` (and `false` and `true`) also all have one byte tuple codes, if we like those better for some reason.

All of this is a 1 byte (per key) thing, and I'm not sure it matters as much on RocksDB as it did on sqlite, as RocksDB compression _should_ be able to compress away common prefixes anyway.